text/0000-pub-sub.md

```diff
@@ -0,0 +1,170 @@
+
+- Time to live (TTL) must be set up for each event sent. It optionally allows a publisher to provide a callback endpoint or an error “Event Type” to be notified that despite the fact that there are subscribers for such an event type no one has received the event within the specified period of time. If a TTL is not explicitly specified for an event, the default TTL for the event type will be used.
+
+- Cross tenant events exchange is assumed in the future, but not in the scope of MVP (v1).
+
+- Integration with external/third party systems should be allowed (not in scope of MVP, but the basis for that must be provided)
+
+### Target Solution Architecture
+
+#### Solution Overview
+- A new module is introduced, `mod-pubsub`. It is responsible for maintaining the registrations of event types and for coordinating distribution of events to the appropriate subscribers. Folio modules may publish events and/or subscribe to events of specific event types. It is the responsibility of publisher modules to define the event types that they will provide, which is accomplished through the use of an event descriptor. Note that it is possible for multiple publishers to register for the same event type. However, the event type may only be defined once. Subscribing modules will register themselves to specific event types by providing a URI that will be invoked by the event manager when pushing events of matching event types.  Event subscription is done on a per tenant basis.
+
+- The `mod-pubsub` module is also responsible for maintaining a trail of activities: event publication; subscription distribution; errors; non-delivery; etc. By using the correlationID embedded in events it is possible to reconstruct the sequence of operations that form a particular workflow.
+
+- By configuring specific event subscriptions for each tenant it is possible to implement specific cases of inter-app coordination.
+
+- The `mod-pubsub` module is implemented using a message queue (Kafka) which will provide the necessary persistence.
+
+#### High-level solution structure
+There is a global FOLIO interface `pubsub` defined in the scope of this solution which provides APIs to ensure the functioning of the subsystem:
+- API for event type registration by publishers and subscribers. This API must be called by back-end modules.
+  - `POST /pubsub/eventtypes` - to register an event type 
+  - `DELETE /pubsub/eventtypes` - to remove event type registration
+  - `POST /pubsub/eventtypes/declare/publisher` - to declare a publisher for a set of event types
+  - `DELETE /pubsub/eventtypes/declare/publisher` - to remove publisher declaration
+  - `POST /pubsub/eventtypes/declare/subscriber` - to declare a subscriber for a set of event types
+  - `DELETE /pubsub/eventtypes/declare/subscriber` - to remove subscriber declaration
+- API to retrieve existing registered event types.
+  - `GET /pubsub/eventtypes` - returns all registered event types
```

Comment:
Is there a way to get details of a specific event type?  Will querying be supported or do you always get "all" registered event types?

Reply:
Thanks for pointing this out. I added such an endpoint.